Error: ERROR_CREATE_NAMESPACE_CONFIG: 2 errors occurred: SetNamespaceReplicationClusters code: 500

<!---
Please note the following potential times when an issue might be in Terraform core:

* [Configuration Language](https://www.terraform.io/docs/configuration/index.html) or resource ordering issues
* [State](https://www.terraform.io/docs/state/index.html) and [State Backend](https://www.terraform.io/docs/backends/index.html) issues
* [Provisioner](https://www.terraform.io/docs/provisioners/index.html) issues
* [Registry](https://registry.terraform.io/) issues
* Spans resources across multiple providers

If you are running into one of these scenarios, we recommend opening an issue in the [Terraform core repository](https://github.com/hashicorp/terraform/) instead.
--->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform Version

<!--- Please run `terraform -v` to show the Terraform core version and provider version(s). If you are not running the latest version of Terraform or the provider, please upgrade because your issue may have already been fixed. [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions). --->
0.12.24

### Affected Resource(s)

<!--- Please list the affected resources and data sources. --->

* pulsar_namespace

### Terraform Configuration Files

<!--- Information about code formatting: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code --->

```hcl
provider "pulsar" {
  web_service_url = "http://localhost:8080"
}

resource "pulsar_cluster" "my_cluster" {
  cluster = "my_cluster"

  cluster_data {
    web_service_url    = "http://broker.vivy.pulsar:8080"
    broker_service_url = "pulsar://broker.vivy.pulsar:6650"
  }
}

resource "pulsar_tenant" "my" {
  tenant           = "my"
  allowed_clusters = ["standalone", pulsar_cluster.my_cluster.cluster]
}

resource "pulsar_namespace" "events" {
  tenant    = pulsar_tenant.my.tenant
  namespace = "events"

  retention_policies {
    retention_minutes    = -1
    retention_size_in_mb = -1
  }
}
```

### Debug Output

<!---
Please provide a link to a GitHub Gist containing the complete debug output. Please do NOT paste the debug output in the issue; just paste a link to the Gist.

To obtain the debug output, see the [Terraform documentation on debugging](https://www.terraform.io/docs/internals/debugging.html).
--->

```
pulsar_cluster.my_cluster: Creating...
pulsar_cluster.my_cluster: Creation complete after 0s [id=my_cluster]
pulsar_tenant.my: Creating...
pulsar_tenant.my: Creation complete after 0s [id=my]
pulsar_namespace.events: Creating...

Error: ERROR_CREATE_NAMESPACE_CONFIG: 2 errors occurred:
        * SetNamespaceReplicationClusters: code: 500 reason: Unknown pulsar error
        * SetRetention: Post "http://localhost:8080/admin/v2/namespaces/my/events/retention": EOF



  on pulsar.tf line 19, in resource "pulsar_namespace" "events":
  19: resource "pulsar_namespace" "events" {
```

In the pulsar log:

```java
16:38:32.400 [pulsar-web-57-16] WARN  org.eclipse.jetty.server.HttpChannel - /admin/v2/namespaces/my/events/replication
javax.servlet.ServletException: javax.servlet.ServletException: java.lang.NullPointerException
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:162) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:173) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.Server.handle(Server.java:494) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:374) [org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:268) [org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311) [org.eclipse.jetty-jetty-io-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103) [org.eclipse.jetty-jetty-io-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117) [org.eclipse.jetty-jetty-io-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336) [org.eclipse.jetty-jetty-util-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313) [org.eclipse.jetty-jetty-util-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171) [org.eclipse.jetty-jetty-util-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129) [org.eclipse.jetty-jetty-util-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:367) [org.eclipse.jetty-jetty-util-9.4.20.v20190813.jar:9.4.20.v20190813]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_232]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_232]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [io.netty-netty-common-4.1.43.Final.jar:4.1.43.Final]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_232]
Caused by: javax.servlet.ServletException: java.lang.NullPointerException
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:432) ~[org.glassfish.jersey.containers-jersey-container-servlet-core-2.27.jar:?]
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:370) ~[org.glassfish.jersey.containers-jersey-container-servlet-core-2.27.jar:?]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:389) ~[org.glassfish.jersey.containers-jersey-container-servlet-core-2.27.jar:?]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:342) ~[org.glassfish.jersey.containers-jersey-container-servlet-core-2.27.jar:?]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:229) ~[org.glassfish.jersey.containers-jersey-container-servlet-core-2.27.jar:?]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:852) ~[org.eclipse.jetty-jetty-servlet-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604) ~[org.eclipse.jetty-jetty-servlet-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.apache.pulsar.broker.web.ResponseHandlerFilter.doFilter(ResponseHandlerFilter.java:53) ~[org.apache.pulsar-pulsar-broker-2.5.0.jar:2.5.0]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1591) ~[org.eclipse.jetty-jetty-servlet-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:542) ~[org.eclipse.jetty-jetty-servlet-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1581) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1307) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:482) ~[org.eclipse.jetty-jetty-servlet-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1549) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1204) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:221) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	... 17 more
Caused by: java.lang.NullPointerException
	at com.google.common.collect.Sets.newHashSet(Sets.java:259) ~[com.google.guava-guava-25.1-jre.jar:?]
	at org.apache.pulsar.broker.admin.impl.NamespacesBase.internalSetNamespaceReplicationClusters(NamespacesBase.java:462) ~[org.apache.pulsar-pulsar-broker-2.5.0.jar:2.5.0]
	at org.apache.pulsar.broker.admin.v2.Namespaces.setNamespaceReplicationClusters(Namespaces.java:254) ~[org.apache.pulsar-pulsar-broker-2.5.0.jar:2.5.0]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_232]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_232]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_232]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_232]
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:76) ~[org.glassfish.jersey.core-jersey-server-2.27.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:148) ~[org.glassfish.jersey.core-jersey-server-2.27.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:191) ~[org.glassfish.jersey.core-jersey-server-2.27.jar:?]
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$VoidOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:183) ~[org.glassfish.jersey.core-jersey-server-2.27.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:103) ~[org.glassfish.jersey.core-jersey-server-2.27.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:493) ~[org.glassfish.jersey.core-jersey-server-2.27.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:415) ~[org.glassfish.jersey.core-jersey-server-2.27.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:104) ~[org.glassfish.jersey.core-jersey-server-2.27.jar:?]
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:277) ~[org.glassfish.jersey.core-jersey-server-2.27.jar:?]
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:272) ~[org.glassfish.jersey.core-jersey-common-2.27.jar:?]
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:268) ~[org.glassfish.jersey.core-jersey-common-2.27.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:316) ~[org.glassfish.jersey.core-jersey-common-2.27.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:298) ~[org.glassfish.jersey.core-jersey-common-2.27.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:268) ~[org.glassfish.jersey.core-jersey-common-2.27.jar:?]
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:289) ~[org.glassfish.jersey.core-jersey-common-2.27.jar:?]
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:256) ~[org.glassfish.jersey.core-jersey-server-2.27.jar:?]
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:703) ~[org.glassfish.jersey.core-jersey-server-2.27.jar:?]
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:416) ~[org.glassfish.jersey.containers-jersey-container-servlet-core-2.27.jar:?]
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:370) ~[org.glassfish.jersey.containers-jersey-container-servlet-core-2.27.jar:?]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:389) ~[org.glassfish.jersey.containers-jersey-container-servlet-core-2.27.jar:?]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:342) ~[org.glassfish.jersey.containers-jersey-container-servlet-core-2.27.jar:?]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:229) ~[org.glassfish.jersey.containers-jersey-container-servlet-core-2.27.jar:?]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:852) ~[org.eclipse.jetty-jetty-servlet-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1604) ~[org.eclipse.jetty-jetty-servlet-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.apache.pulsar.broker.web.ResponseHandlerFilter.doFilter(ResponseHandlerFilter.java:53) ~[org.apache.pulsar-pulsar-broker-2.5.0.jar:2.5.0]
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1591) ~[org.eclipse.jetty-jetty-servlet-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:542) ~[org.eclipse.jetty-jetty-servlet-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1581) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1307) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:482) ~[org.eclipse.jetty-jetty-servlet-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1549) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1204) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:221) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146) ~[org.eclipse.jetty-jetty-server-9.4.20.v20190813.jar:9.4.20.v20190813]
	... 17 more
16:38:32.407 [pulsar-web-57-16] INFO  org.eclipse.jetty.server.RequestLog - 172.17.0.1 - - [02/Apr/2020:16:38:32 +0000] "POST /admin/v2/namespaces/my/events/replication HTTP/1.1" 500 0 "-" "None" 17
```
### Panic Output

<!--- If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`. --->

### Expected Behavior

No errors

### Actual Behavior

Seems when additional namespace configuration is not defined, then provider sends empty value and that is not ok for pulsar. 

Seems non-defined configuration should be handled a bit differently

### Steps to Reproduce

<!--- Please list the steps required to reproduce the issue. --->

1. `terraform apply`

### Important Factoids

<!--- Are there anything atypical about your accounts that we should know? For example: Running in EC2 Classic? --->

### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Vendor documentation? For example:
--->


I can try to fix that

--- a/pulsar/resource_pulsar_namespace.go
+++ b/pulsar/resource_pulsar_namespace.go
@@ -21,6 +21,7 @@ import (
 	"bytes"
 	"fmt"
 	"strconv"
+	"strings"
 
 	"github.com/hashicorp/go-multierror"
 
@@ -60,6 +61,7 @@ func resourcePulsarNamespace() *schema.Resource {
 				Type:        schema.TypeSet,
 				Optional:    true,
 				Description: descriptions["dispatch_rate"],
+				MaxItems:    1,
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
 						"dispatch_msg_throttling_rate": {
@@ -81,6 +83,7 @@ func resourcePulsarNamespace() *schema.Resource {
 			"retention_policies": {
 				Type:     schema.TypeSet,
 				Optional: true,
+				MaxItems: 1,
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
 						"retention_minutes": {
@@ -99,27 +102,60 @@ func resourcePulsarNamespace() *schema.Resource {
 				Type:        schema.TypeSet,
 				Optional:    true,
 				Description: descriptions["namespace_config"],
+				MaxItems:    1,
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
 						"anti_affinity": {
 							Type:     schema.TypeString,
 							Optional: true,
+							ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
+								v := val.(string)
+								if len(strings.Trim(strings.TrimSpace(v), "\"")) == 0 {
+									errs = append(errs, fmt.Errorf("%q must not be empty", key))
+								}
+								return
+							},
 						},
 						"max_consumers_per_subscription": {
 							Type:     schema.TypeInt,
 							Optional: true,
+							Default:  -1,
+							ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
+								v := val.(int)
+								if v < 0 {
+									errs = append(errs, fmt.Errorf("%q must be 0 or more, got: %d", key, v))
+								}
+								return
+							},
 						},
 						"max_consumers_per_topic": {
 							Type:     schema.TypeInt,
 							Optional: true,
+							Default:  -1,
+							ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
+								v := val.(int)
+								if v < 0 {
+									errs = append(errs, fmt.Errorf("%q must be 0 or more, got: %d", key, v))
+								}
+								return
+							},
 						},
 						"max_producers_per_topic": {
 							Type:     schema.TypeInt,
 							Optional: true,
+							Default:  -1,
+							ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
+								v := val.(int)
+								if v < 0 {
+									errs = append(errs, fmt.Errorf("%q must be 0 or more, got: %d", key, v))
+								}
+								return
+							},
 						},
 						"replication_clusters": {
 							Type:     schema.TypeList,
 							Optional: true,
+							MinItems: 1,
 							Elem: &schema.Schema{
 								Type: schema.TypeString,
 							},
@@ -131,6 +167,7 @@ func resourcePulsarNamespace() *schema.Resource {
 			"persistence_policies": {
 				Type:     schema.TypeSet,
 				Optional: true,
+				MaxItems: 1,
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
 						"bookkeeper_ensemble": {
@@ -178,16 +215,11 @@ func resourcePulsarNamespaceCreate(d *schema.ResourceData, meta interface{}) err
 		return fmt.Errorf("ERROR_CREATE_NAMESPACE: %w", err)
 	}
 
-	enableDeduplication := d.Get("enable_deduplication").(bool)
+	enableDeduplication, deduplicationDefined := d.GetOk("enable_deduplication")
 	namespaceConfig := d.Get("namespace_config").(*schema.Set)
 	retentionPoliciesConfig := d.Get("retention_policies").(*schema.Set)
 	dispatchRateConfig := d.Get("dispatch_rate").(*schema.Set)
 	persistencePoliciesConfig := d.Get("persistence_policies").(*schema.Set)
-
-	dispatchRate := unmarshalDispatchRate(dispatchRateConfig)
-	retentionPolicies := unmarshalRetentionPolicies(retentionPoliciesConfig)
-	nsCfg := unmarshalNamespaceConfig(namespaceConfig)
-	persistencePolicies := unmarshalPersistencePolicies(persistencePoliciesConfig)
 
 	nsName, err := utils.GetNameSpaceName(tenant, namespace)
 	if err != nil {
@@ -196,40 +228,68 @@ func resourcePulsarNamespaceCreate(d *schema.ResourceData, meta interface{}) err
 
 	var errs error
 
-	if err = client.SetNamespaceAntiAffinityGroup(nsName.String(), nsCfg.AntiAffinity); err != nil {
-		errs = multierror.Append(errs, fmt.Errorf("SetNamespaceAntiAffinityGroup: %w", err))
+	if namespaceConfig.Len() > 0 {
+		nsCfg := unmarshalNamespaceConfig(namespaceConfig)
+
+		if len(nsCfg.AntiAffinity) > 0 {
+			if err = client.SetNamespaceAntiAffinityGroup(nsName.String(), nsCfg.AntiAffinity); err != nil {
+				errs = multierror.Append(errs, fmt.Errorf("SetNamespaceAntiAffinityGroup: %w", err))
+			}
+		}
+
+		if len(nsCfg.ReplicationClusters) > 0 {
+			if err = client.SetNamespaceReplicationClusters(nsName.String(), nsCfg.ReplicationClusters); err != nil {
+				errs = multierror.Append(errs, fmt.Errorf("SetNamespaceReplicationClusters: %w", err))
+			}
+		}
+
+		if nsCfg.MaxConsumersPerTopic >= 0 {
+			if err = client.SetMaxConsumersPerTopic(*nsName, nsCfg.MaxConsumersPerTopic); err != nil {
+				errs = multierror.Append(errs, fmt.Errorf("SetMaxConsumersPerTopic: %w", err))
+			}
+		}
+
+		if nsCfg.MaxConsumersPerSubscription >= 0 {
+			if err = client.SetMaxConsumersPerSubscription(*nsName, nsCfg.MaxConsumersPerSubscription); err != nil {
+				errs = multierror.Append(errs, fmt.Errorf("SetMaxConsumersPerSubscription: %w", err))
+			}
+		}
+
+		if nsCfg.MaxProducersPerTopic >= 0 {
+			if err = client.SetMaxProducersPerTopic(*nsName, nsCfg.MaxProducersPerTopic); err != nil {
+				errs = multierror.Append(errs, fmt.Errorf("SetMaxProducersPerTopic: %w", err))
+			}
+		}
 	}
 
-	if err = client.SetDispatchRate(*nsName, *dispatchRate); err != nil {
-		errs = multierror.Append(errs, fmt.Errorf("SetDispatchRate: %w", err))
+	if dispatchRateConfig.Len() > 0 {
+		dispatchRate := unmarshalDispatchRate(dispatchRateConfig)
+
+		if err = client.SetDispatchRate(*nsName, *dispatchRate); err != nil {
+			errs = multierror.Append(errs, fmt.Errorf("SetDispatchRate: %w", err))
+		}
 	}
 
-	if err = client.SetNamespaceReplicationClusters(nsName.String(), nsCfg.ReplicationClusters); err != nil {
-		errs = multierror.Append(errs, fmt.Errorf("SetNamespaceReplicationClusters: %w", err))
+	if retentionPoliciesConfig.Len() > 0 {
+		retentionPolicies := unmarshalRetentionPolicies(retentionPoliciesConfig)
+
+		if err = client.SetRetention(nsName.String(), *retentionPolicies); err != nil {
+			errs = multierror.Append(errs, fmt.Errorf("SetRetention: %w", err))
+		}
 	}
 
-	if err = client.SetRetention(nsName.String(), *retentionPolicies); err != nil {
-		errs = multierror.Append(errs, fmt.Errorf("SetRetention: %w", err))
+	if persistencePoliciesConfig.Len() > 0 {
+		persistencePolicies := unmarshalPersistencePolicies(persistencePoliciesConfig)
+
+		if err = client.SetPersistence(nsName.String(), *persistencePolicies); err != nil {
+			errs = multierror.Append(errs, fmt.Errorf("SetPersistence: %w", err))
+		}
 	}
 
-	if err = client.SetMaxConsumersPerTopic(*nsName, nsCfg.MaxConsumersPerTopic); err != nil {
-		errs = multierror.Append(errs, fmt.Errorf("SetMaxConsumersPerTopic: %w", err))
-	}
-
-	if err = client.SetMaxConsumersPerSubscription(*nsName, nsCfg.MaxConsumersPerSubscription); err != nil {
-		errs = multierror.Append(errs, fmt.Errorf("SetMaxConsumersPerSubscript1ion: %w", err))
-	}
-
-	if err = client.SetMaxProducersPerTopic(*nsName, nsCfg.MaxProducersPerTopic); err != nil {
-		errs = multierror.Append(errs, fmt.Errorf("SetMaxProducersPerTopic: %w", err))
-	}
-
-	if err = client.SetPersistence(nsName.String(), *persistencePolicies); err != nil {
-		errs = multierror.Append(errs, fmt.Errorf("SetPersistence: %w", err))
-	}
-
-	if err = client.SetDeduplicationStatus(nsName.String(), enableDeduplication); err != nil {
-		errs = multierror.Append(errs, fmt.Errorf("SetDeduplicationStatus: %w", err))
+	if deduplicationDefined {
+		if err = client.SetDeduplicationStatus(nsName.String(), enableDeduplication.(bool)); err != nil {
+			errs = multierror.Append(errs, fmt.Errorf("SetDeduplicationStatus: %w", err))
+		}
 	}
 
 	if errs != nil {
@@ -243,10 +303,7 @@ func resourcePulsarNamespaceCreate(d *schema.ResourceData, meta interface{}) err
 }
 
 func resourcePulsarNamespaceRead(d *schema.ResourceData, meta interface{}) error {
-
-	// pulsar client is not required as we just need to find the full namespace name,
-	// which can be achieved from utils package from the pulsarctl module
-	_ = meta
+	client := meta.(pulsar.Client).Namespaces()
 
 	tenant := d.Get("tenant").(string)
 	namespace := d.Get("namespace").(string)
@@ -265,6 +322,64 @@ func resourcePulsarNamespaceRead(d *schema.ResourceData, meta interface{}) error
 	_ = d.Set("namespace", namespace)
 	_ = d.Set("tenant", tenant)
 
+	if namespaceConfig, ok := d.GetOk("namespace_config"); ok && namespaceConfig.(*schema.Set).Len() > 0 {
+		afgrp, err := client.GetNamespaceAntiAffinityGroup(ns.String())
+		if err != nil {
+			return fmt.Errorf("ERROR_READ_NAMESPACE: GetNamespaceAntiAffinityGroup: %w", err)
+		}
+
+		maxConsPerSub, err := client.GetMaxConsumersPerSubscription(*ns)
+		if err != nil {
+			return fmt.Errorf("ERROR_READ_NAMESPACE: GetMaxConsumersPerSubscription: %w", err)
+		}
+
+		maxConsPerTopic, err := client.GetMaxConsumersPerTopic(*ns)
+		if err != nil {
+			return fmt.Errorf("ERROR_READ_NAMESPACE: GetMaxConsumersPerTopic: %w", err)
+		}
+
+		maxProdPerTopic, err := client.GetMaxProducersPerTopic(*ns)
+		if err != nil {
+			return fmt.Errorf("ERROR_READ_NAMESPACE: GetMaxProducersPerTopic: %w", err)
+		}
+
+		replClustersRaw, err := client.GetNamespaceReplicationClusters(ns.String())
+		if err != nil {
+			return fmt.Errorf("ERROR_READ_NAMESPACE: GetMaxProducersPerTopic: %w", err)
+		}
+
+		replClusters := make([]interface{}, len(replClustersRaw))
+		for i, cl := range replClustersRaw {
+			replClusters[i] = cl
+		}
+
+		_ = d.Set("namespace_config", schema.NewSet(namespaceConfigToHash, []interface{}{
+			map[string]interface{}{
+				"anti_affinity":                  strings.Trim(strings.TrimSpace(afgrp), "\""),
+				"max_consumers_per_subscription": maxConsPerSub,
+				"max_consumers_per_topic":        maxConsPerTopic,
+				"max_producers_per_topic":        maxProdPerTopic,
+				"replication_clusters":           replClusters,
+			},
+		}))
+	}
+
+	if persistencePoliciesConfig, ok := d.GetOk("persistence_policies"); ok && persistencePoliciesConfig.(*schema.Set).Len() > 0 {
+		persistence, err := client.GetPersistence(ns.String())
+		if err != nil {
+			return fmt.Errorf("ERROR_READ_NAMESPACE: GetPersistence: %w", err)
+		}
+
+		_ = d.Set("persistence_policies", schema.NewSet(persistencePoliciesToHash, []interface{}{
+			map[string]interface{}{
+				"bookkeeper_ensemble":                 persistence.BookkeeperEnsemble,
+				"bookkeeper_write_quorum":             persistence.BookkeeperWriteQuorum,
+				"bookkeeper_ack_quorum":               persistence.BookkeeperAckQuorum,
+				"managed_ledger_max_mark_delete_rate": persistence.ManagedLedgerMaxMarkDeleteRate,
+			},
+		}))
+	}
+
 	return nil
 }
 
@@ -273,16 +388,11 @@ func resourcePulsarNamespaceUpdate(d *schema.ResourceData, meta interface{}) err
 
 	namespace := d.Get("namespace").(string)
 	tenant := d.Get("tenant").(string)
-	enableDeduplication := d.Get("enable_deduplication").(bool)
+	enableDeduplication, deduplicationDefined := d.GetOk("enable_deduplication")
 	namespaceConfig := d.Get("namespace_config").(*schema.Set)
 	retentionPoliciesConfig := d.Get("retention_policies").(*schema.Set)
 	dispatchRateConfig := d.Get("dispatch_rate").(*schema.Set)
 	persistencePoliciesConfig := d.Get("persistence_policies").(*schema.Set)
-
-	nsCfg := unmarshalNamespaceConfig(namespaceConfig)
-	dispatchRate := unmarshalDispatchRate(dispatchRateConfig)
-	retentionPolicies := unmarshalRetentionPolicies(retentionPoliciesConfig)
-	persistencePolicies := unmarshalPersistencePolicies(persistencePoliciesConfig)
 
 	nsName, err := utils.GetNameSpaceName(tenant, namespace)
 	if err != nil {
@@ -291,15 +401,66 @@ func resourcePulsarNamespaceUpdate(d *schema.ResourceData, meta interface{}) err
 
 	var errs error
 
-	errs = multierror.Append(errs, client.SetNamespaceAntiAffinityGroup(nsName.String(), nsCfg.AntiAffinity))
-	errs = multierror.Append(errs, client.SetDispatchRate(*nsName, *dispatchRate))
-	errs = multierror.Append(errs, client.SetNamespaceReplicationClusters(nsName.String(), nsCfg.ReplicationClusters))
-	errs = multierror.Append(errs, client.SetRetention(nsName.String(), *retentionPolicies))
-	errs = multierror.Append(errs, client.SetMaxConsumersPerTopic(*nsName, nsCfg.MaxConsumersPerTopic))
-	errs = multierror.Append(errs, client.SetMaxConsumersPerSubscription(*nsName, nsCfg.MaxConsumersPerSubscription))
-	errs = multierror.Append(errs, client.SetMaxProducersPerTopic(*nsName, nsCfg.MaxProducersPerTopic))
-	errs = multierror.Append(errs, client.SetPersistence(nsName.String(), *persistencePolicies))
-	errs = multierror.Append(errs, client.SetDeduplicationStatus(nsName.String(), enableDeduplication))
+	if namespaceConfig.Len() > 0 {
+		nsCfg := unmarshalNamespaceConfig(namespaceConfig)
+
+		if len(nsCfg.AntiAffinity) > 0 {
+			if err = client.SetNamespaceAntiAffinityGroup(nsName.String(), nsCfg.AntiAffinity); err != nil {
+				errs = multierror.Append(errs, fmt.Errorf("SetNamespaceAntiAffinityGroup: %w", err))
+			}
+		}
+
+		if len(nsCfg.ReplicationClusters) > 0 {
+			if err = client.SetNamespaceReplicationClusters(nsName.String(), nsCfg.ReplicationClusters); err != nil {
+				errs = multierror.Append(errs, fmt.Errorf("SetNamespaceReplicationClusters: %w", err))
+			}
+		}
+
+		if nsCfg.MaxConsumersPerTopic >= 0 {
+			if err = client.SetMaxConsumersPerTopic(*nsName, nsCfg.MaxConsumersPerTopic); err != nil {
+				errs = multierror.Append(errs, fmt.Errorf("SetMaxConsumersPerTopic: %w", err))
+			}
+		}
+
+		if nsCfg.MaxConsumersPerSubscription >= 0 {
+			if err = client.SetMaxConsumersPerSubscription(*nsName, nsCfg.MaxConsumersPerSubscription); err != nil {
+				errs = multierror.Append(errs, fmt.Errorf("SetMaxConsumersPerSubscription: %w", err))
+			}
+		}
+
+		if nsCfg.MaxProducersPerTopic >= 0 {
+			if err = client.SetMaxProducersPerTopic(*nsName, nsCfg.MaxProducersPerTopic); err != nil {
+				errs = multierror.Append(errs, fmt.Errorf("SetMaxProducersPerTopic: %w", err))
+			}
+		}
+	}
+
+	if retentionPoliciesConfig.Len() > 0 {
+		retentionPolicies := unmarshalRetentionPolicies(retentionPoliciesConfig)
+		if err = client.SetRetention(nsName.String(), *retentionPolicies); err != nil {
+			errs = multierror.Append(errs, fmt.Errorf("SetRetention: %w", err))
+		}
+	}
+
+	if dispatchRateConfig.Len() > 0 {
+		dispatchRate := unmarshalDispatchRate(dispatchRateConfig)
+		if err = client.SetDispatchRate(*nsName, *dispatchRate); err != nil {
+			errs = multierror.Append(errs, fmt.Errorf("SetDispatchRate: %w", err))
+		}
+	}
+
+	if persistencePoliciesConfig.Len() > 0 {
+		persistencePolicies := unmarshalPersistencePolicies(persistencePoliciesConfig)
+		if err = client.SetPersistence(nsName.String(), *persistencePolicies); err != nil {
+			errs = multierror.Append(errs, fmt.Errorf("SetPersistence: %w", err))
+		}
+	}
+
+	if deduplicationDefined {
+		if err = client.SetDeduplicationStatus(nsName.String(), enableDeduplication.(bool)); err != nil {
+			errs = multierror.Append(errs, fmt.Errorf("SetDeduplicationStatus: %w", err))
+		}
+	}
 
 	if errs != nil {
 		return fmt.Errorf("ERROR_UPDATE_NAMESPACE_CONFIG: %w", errs)
@@ -342,7 +503,7 @@ func resourcePulsarNamespaceExists(d *schema.ResourceData, meta interface{}) (bo
 
 	nsList, err := client.GetNamespaces(tenant)
 	if err != nil {
-		return false, fmt.Errorf("ERROR_READING_NAMESPACE_NAME: %w", err)
+		return false, fmt.Errorf("ERROR_READING_NAMESPACES: %w", err)
 	}
 
 	for _, ns := range nsList {

--- a/pulsar/resource_pulsar_namespace.go
+++ b/pulsar/resource_pulsar_namespace.go
@@ -364,7 +364,7 @@ func resourcePulsarNamespaceRead(d *schema.ResourceData, meta interface{}) error
 		}))
 	}
 
-	if persistencePoliciesConfig, ok := d.GetOk("persistence_policies"); ok && persistencePoliciesConfig.(*schema.Set).Len() > 0 {
+	if persPoliciesCfg, ok := d.GetOk("persistence_policies"); ok && persPoliciesCfg.(*schema.Set).Len() > 0 {
 		persistence, err := client.GetPersistence(ns.String())
 		if err != nil {
 			return fmt.Errorf("ERROR_READ_NAMESPACE: GetPersistence: %w", err)

--- a/pulsar/resource_pulsar_namespace_test.go
+++ b/pulsar/resource_pulsar_namespace_test.go
@@ -167,6 +167,7 @@ func TestNamespaceWithUndefinedOptionalsUpdate(t *testing.T) {
 	})
 }
 
+//nolint:unparam
 func testPulsarNamespaceExists(ns string) resource.TestCheckFunc {
 	return func(s *terraform.State) error {
 		rs, ok := s.RootModule().Resources[ns]

--- a/pulsar/resource_pulsar_namespace_test.go
+++ b/pulsar/resource_pulsar_namespace_test.go
@@ -19,9 +19,10 @@ package pulsar
 
 import (
 	"fmt"
-	"github.com/hashicorp/terraform-plugin-sdk/helper/acctest"
 	"strings"
 	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/acctest"
 
 	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/terraform"

--- a/pulsar/resource_pulsar_namespace_test.go
+++ b/pulsar/resource_pulsar_namespace_test.go
@@ -19,6 +19,7 @@ package pulsar
 
 import (
 	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/acctest"
 	"strings"
 	"testing"
 
@@ -71,6 +72,9 @@ func testSweepNS(url string) error {
 func TestNamespace(t *testing.T) {
 
 	resourceName := "pulsar_namespace.test"
+	cName := acctest.RandString(10)
+	tName := acctest.RandString(10)
+	nsName := acctest.RandString(10)
 
 	resource.Test(t, resource.TestCase{
 		PreCheck:      func() { testAccPreCheck(t) },
@@ -79,9 +83,46 @@ func TestNamespace(t *testing.T) {
 		CheckDestroy:  testPulsarNamespaceDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testPulsarNamespace,
+				Config: testPulsarNamespace(testWebServiceURL, cName, tName, nsName),
 				Check: resource.ComposeTestCheckFunc(
 					testPulsarNamespaceExists(resourceName),
+				),
+			},
+		},
+	})
+}
+
+func TestNamespaceWithUpdate(t *testing.T) {
+
+	resourceName := "pulsar_namespace.test"
+	cName := acctest.RandString(10)
+	tName := acctest.RandString(10)
+	nsName := acctest.RandString(10)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:      func() { testAccPreCheck(t) },
+		Providers:     testAccProviders,
+		IDRefreshName: resourceName,
+		CheckDestroy:  testPulsarNamespaceDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testPulsarNamespaceWithoutOptionals(testWebServiceURL, cName, tName, nsName),
+				Check: resource.ComposeTestCheckFunc(
+					testPulsarNamespaceExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "dispatch_rate.#", "0"),
+					resource.TestCheckResourceAttr(resourceName, "retention_policies.#", "0"),
+					resource.TestCheckResourceAttr(resourceName, "namespace_config.#", "0"),
+					resource.TestCheckNoResourceAttr(resourceName, "enable_deduplication"),
+				),
+			},
+			{
+				Config: testPulsarNamespace(testWebServiceURL, cName, tName, nsName),
+				Check: resource.ComposeTestCheckFunc(
+					testPulsarNamespaceExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "dispatch_rate.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "retention_policies.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "namespace_config.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "enable_deduplication", "true"),
 				),
 			},
 		},
@@ -117,7 +158,7 @@ func testPulsarNamespaceExists(ns string) resource.TestCheckFunc {
 			}
 		}
 
-		return fmt.Errorf("ERROR_RESOURCE_NAMESPACE_DOES_NOT_EXISTS")
+		return fmt.Errorf(`ERROR_RESOURCE_NAMESPACE_DOES_NOT_EXISTS: "%s"`, ns)
 	}
 }
 
@@ -152,14 +193,14 @@ func testPulsarNamespaceDestroy(s *terraform.State) error {
 	return nil
 }
 
-var (
-	testPulsarNamespace = fmt.Sprintf(`
+func testPulsarNamespaceWithoutOptionals(wsUrl, cluster, tenant, ns string) string {
+	return fmt.Sprintf(`
 provider "pulsar" {
   web_service_url = "%s"
 }
 
 resource "pulsar_cluster" "test_cluster" {
-  cluster = "skrulls"
+  cluster = "%s"
 
   cluster_data {
     web_service_url    = "http://localhost:8080"
@@ -170,13 +211,42 @@ resource "pulsar_cluster" "test_cluster" {
 }
 
 resource "pulsar_tenant" "test_tenant" {
-  tenant           = "thanos"
+  tenant           = "%s"
   allowed_clusters = [pulsar_cluster.test_cluster.cluster, "standalone"]
 }
 
 resource "pulsar_namespace" "test" {
   tenant    = pulsar_tenant.test_tenant.tenant
-  namespace = "eternals"
+  namespace = "%s"
+}
+`, wsUrl, cluster, tenant, ns)
+}
+
+func testPulsarNamespace(wsUrl, cluster, tenant, ns string) string {
+	return fmt.Sprintf(`
+provider "pulsar" {
+  web_service_url = "%s"
+}
+
+resource "pulsar_cluster" "test_cluster" {
+  cluster = "%s"
+
+  cluster_data {
+    web_service_url    = "http://localhost:8080"
+    broker_service_url = "http://localhost:6050"
+    peer_clusters      = ["standalone"]
+  }
+
+}
+
+resource "pulsar_tenant" "test_tenant" {
+  tenant           = "%s"
+  allowed_clusters = [pulsar_cluster.test_cluster.cluster, "standalone"]
+}
+
+resource "pulsar_namespace" "test" {
+  tenant    = pulsar_tenant.test_tenant.tenant
+  namespace = "%s"
 
   enable_deduplication = true
 
@@ -200,12 +270,12 @@ resource "pulsar_namespace" "test" {
   }
 
   persistence_policies {
-    bookkeeper_ensemble                   = 1
-    bookkeeper_write_quorum               = 1
-    bookkeeper_ack_quorum                 = 1
+    bookkeeper_ensemble                   = 2
+    bookkeeper_write_quorum               = 2
+    bookkeeper_ack_quorum                 = 2
     managed_ledger_max_mark_delete_rate   = 0.0
   }
 
 }
-`, testWebServiceURL)
-)
+`, wsUrl, cluster, tenant, ns)
+}

--- a/pulsar/resource_pulsar_namespace_test.go
+++ b/pulsar/resource_pulsar_namespace_test.go
@@ -129,6 +129,44 @@ func TestNamespaceWithUpdate(t *testing.T) {
 	})
 }
 
+func TestNamespaceWithUndefinedOptionalsUpdate(t *testing.T) {
+
+	resourceName := "pulsar_namespace.test"
+	cName := acctest.RandString(10)
+	tName := acctest.RandString(10)
+	nsName := acctest.RandString(10)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:      func() { testAccPreCheck(t) },
+		Providers:     testAccProviders,
+		IDRefreshName: resourceName,
+		CheckDestroy:  testPulsarNamespaceDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testPulsarNamespaceWithoutOptionals(testWebServiceURL, cName, tName, nsName),
+				Check: resource.ComposeTestCheckFunc(
+					testPulsarNamespaceExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "dispatch_rate.#", "0"),
+					resource.TestCheckResourceAttr(resourceName, "retention_policies.#", "0"),
+					resource.TestCheckResourceAttr(resourceName, "namespace_config.#", "0"),
+					resource.TestCheckNoResourceAttr(resourceName, "enable_deduplication"),
+				),
+			},
+			{
+				Config: testPulsarNamespaceWithUndefinedOptionalsInNsConf(testWebServiceURL, cName, tName, nsName),
+				Check: resource.ComposeTestCheckFunc(
+					testPulsarNamespaceExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "dispatch_rate.#", "0"),
+					resource.TestCheckResourceAttr(resourceName, "retention_policies.#", "0"),
+					resource.TestCheckResourceAttr(resourceName, "namespace_config.#", "1"),
+					resource.TestCheckNoResourceAttr(resourceName, "enable_deduplication"),
+				),
+				ExpectNonEmptyPlan: true,
+			},
+		},
+	})
+}
+
 func testPulsarNamespaceExists(ns string) resource.TestCheckFunc {
 	return func(s *terraform.State) error {
 		rs, ok := s.RootModule().Resources[ns]
@@ -193,7 +231,7 @@ func testPulsarNamespaceDestroy(s *terraform.State) error {
 	return nil
 }
 
-func testPulsarNamespaceWithoutOptionals(wsUrl, cluster, tenant, ns string) string {
+func testPulsarNamespaceWithoutOptionals(wsURL, cluster, tenant, ns string) string {
 	return fmt.Sprintf(`
 provider "pulsar" {
   web_service_url = "%s"
@@ -219,10 +257,10 @@ resource "pulsar_namespace" "test" {
   tenant    = pulsar_tenant.test_tenant.tenant
   namespace = "%s"
 }
-`, wsUrl, cluster, tenant, ns)
+`, wsURL, cluster, tenant, ns)
 }
 
-func testPulsarNamespace(wsUrl, cluster, tenant, ns string) string {
+func testPulsarNamespace(wsURL, cluster, tenant, ns string) string {
 	return fmt.Sprintf(`
 provider "pulsar" {
   web_service_url = "%s"
@@ -277,5 +315,40 @@ resource "pulsar_namespace" "test" {
   }
 
 }
-`, wsUrl, cluster, tenant, ns)
+`, wsURL, cluster, tenant, ns)
+}
+
+func testPulsarNamespaceWithUndefinedOptionalsInNsConf(wsURL, cluster, tenant, ns string) string {
+	return fmt.Sprintf(`
+provider "pulsar" {
+  web_service_url = "%s"
+}
+
+resource "pulsar_cluster" "test_cluster" {
+  cluster = "%s"
+
+  cluster_data {
+    web_service_url    = "http://localhost:8080"
+    broker_service_url = "http://localhost:6050"
+    peer_clusters      = ["standalone"]
+  }
+
+}
+
+resource "pulsar_tenant" "test_tenant" {
+  tenant           = "%s"
+  allowed_clusters = [pulsar_cluster.test_cluster.cluster, "standalone"]
+}
+
+resource "pulsar_namespace" "test" {
+  tenant    = pulsar_tenant.test_tenant.tenant
+  namespace = "%s"
+
+  namespace_config {
+    anti_affinity                  = "anti-aff"
+    max_producers_per_topic        = "50"
+  }
+
+}
+`, wsURL, cluster, tenant, ns)
 }

--- a/pulsar/resource_pulsar_tenant.go
+++ b/pulsar/resource_pulsar_tenant.go
@@ -19,6 +19,7 @@ package pulsar
 
 import (
 	"fmt"
+	"sort"
 	"strings"
 
 	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"

--- a/pulsar/resource_pulsar_tenant.go
+++ b/pulsar/resource_pulsar_tenant.go
@@ -19,7 +19,6 @@ package pulsar
 
 import (
 	"fmt"
-	"sort"
 	"strings"
 
 	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
@@ -41,7 +40,7 @@ func resourcePulsarTenant() *schema.Resource {
 				Description: descriptions["tenant"],
 			},
 			"allowed_clusters": {
-				Type:        schema.TypeList,
+				Type:        schema.TypeSet,
 				Optional:    true,
 				Description: descriptions["allowed_clusters"],
 				Elem:        &schema.Schema{Type: schema.TypeString},
@@ -61,7 +60,7 @@ func resourcePulsarTenantCreate(d *schema.ResourceData, meta interface{}) error 
 
 	tenant := d.Get("tenant").(string)
 	adminRoles := handleHCLArray(d, "admin_roles")
-	allowedClusters := handleHCLArray(d, "allowed_clusters")
+	allowedClusters := handleHCLArrayV2(d.Get("allowed_clusters").(*schema.Set).List())
 
 	input := utils.TenantData{
 		Name:            tenant,
@@ -86,10 +85,6 @@ func resourcePulsarTenantRead(d *schema.ResourceData, meta interface{}) error {
 		return fmt.Errorf("ERROR_READ_TENANT: %w", err)
 	}
 
-	// pulsar sometimes returns it in random order
-	sort.Strings(td.AdminRoles)
-	sort.Strings(td.AllowedClusters)
-
 	_ = d.Set("tenant", tenant)
 	_ = d.Set("admin_roles", td.AdminRoles)
 	_ = d.Set("allowed_clusters", td.AllowedClusters)
@@ -104,7 +99,7 @@ func resourcePulsarTenantUpdate(d *schema.ResourceData, meta interface{}) error 
 	d.Partial(true)
 	tenant := d.Get("tenant").(string)
 	adminRoles := handleHCLArray(d, "admin_roles")
-	allowedClusters := handleHCLArray(d, "allowed_clusters")
+	allowedClusters := handleHCLArrayV2(d.Get("allowed_clusters").(*schema.Set).List())
 
 	input := utils.TenantData{
 		Name:            tenant,

--- a/pulsar/resource_pulsar_tenant.go
+++ b/pulsar/resource_pulsar_tenant.go
@@ -85,6 +85,10 @@ func resourcePulsarTenantRead(d *schema.ResourceData, meta interface{}) error {
 		return fmt.Errorf("ERROR_READ_TENANT: %w", err)
 	}
 
+	// pulsar sometimes returns it in random order
+	sort.Strings(td.AdminRoles)
+	sort.Strings(td.AllowedClusters)
+
 	_ = d.Set("tenant", tenant)
 	_ = d.Set("admin_roles", td.AdminRoles)
 	_ = d.Set("allowed_clusters", td.AllowedClusters)
@@ -163,17 +167,7 @@ func deleteExistingNamespacesForTenant(tenant string, meta interface{}) error {
 
 func handleHCLArray(d *schema.ResourceData, key string) []string {
 	hclArray := d.Get(key).([]interface{})
-	out := make([]string, 0)
-
-	if len(hclArray) == 0 {
-		return out
-	}
-
-	for _, value := range hclArray {
-		out = append(out, value.(string))
-	}
-
-	return out
+	return handleHCLArrayV2(hclArray)
 }
 
 func handleHCLArrayV2(hclArray []interface{}) []string {

